src/library/scala/concurrent/Future.scala

```diff
@@ -628,17 +698,20 @@ object Future {
    * @param futures   the `IterableOnce` of Futures in which to find the first completed
    * @return          the `Future` holding the result of the future that is first to be completed
    */
-  def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] = {
-    val p = Promise[T]()
-    val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
-      override def apply(v1: Try[T]): Unit = getAndSet(null) match {
-        case null => ()
-        case some => some tryComplete v1
+  final def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] =
+    if (futures.isEmpty) Future.failed[T](Future.firstCompletedOfException)
+    else {
+      val p = Promise[T]()
+      val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
+        override def apply(v1: Try[T]): Unit =  {
+          val r = getAndSet(null)
+          if (r ne null)
+            r tryComplete v1
```

Comment:
I agree, seems like there are conflicting ideas here. if using `tryComplete` then the code could be simplified to.

`futures.iterator().foreach { _.onComplete(p.tryComplete) }`

since promises are atomic refs anyway.

Reply:
Note that the AtomicReference is not new—see History for background.
tryComplete is used because it is cheaper.

Reply:
Aha. An inline comment "// cheaper than \`complete\`" might be helpful

Reply:
@NthPortal complete is tryComplete + conditional throw